minerl/Malmo/Minecraft/src/main/java/com/microsoft/Malmo/MissionHandlers/ObservationFromDamageImplementation.java

```diff
@@ -0,0 +1,142 @@
+import net.minecraftforge.event.entity.living.LivingHurtEvent;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+
+import static java.lang.Math.asin;
+import static java.lang.Math.atan2;
+
+/**
+ * Simple IObservationProducer object that pings out a whole bunch of data.<br>
+ */
+public class ObservationFromDamageImplementation extends HandlerBase implements IObservationProducer {
+    private DamageSource damageSource;
+    private float damageAmount;
+    private EntityLivingBase entity;
+    private boolean hasDied = false;
+    private boolean damageSourceReplaced = false;
+
+
+    @Override
+    public void prepare(MissionInit missionInit) {
+        MinecraftForge.EVENT_BUS.register(this);
+    }
+
+    @Override
+    public void cleanup() {
+    }
+
+    @SubscribeEvent
+    public void onDeath(LivingDeathEvent event) {
+        if (event.getEntityLiving().equals(Minecraft.getMinecraft().player)) {
+            if (this.damageSource != null) {
+                // If the agent dies from damage, a LivingHurtEvent will precede the death event. Since death events do
+                // not include the damage amount we simply update the other fields to ensure the cause of death is
+                // properly reflected
+                this.damageSourceReplaced = true;
+            }
+            this.damageSource = event.getSource();
+            this.entity = event.getEntityLiving();
+            this.hasDied = true;
+        }
+    }
+
+    @SubscribeEvent
+    public void onDamage(LivingHurtEvent event) {
+        if (event.getEntityLiving().equals(Minecraft.getMinecraft().player)) {
+            if (this.damageSource != null) {
+                this.damageSourceReplaced = true;
+                // If multiple LivingHurtEvents pertain to the player, we keep the most recent one and record the fact
+                // that we are skipping a damage event in leiu of building an array based reporting system
+                System.out.println("Warning skipped damage event - multiple damage events in one tick!");
+            }
+            this.damageSource = event.getSource();
+            this.damageAmount = event.getAmount();
+            this.entity = event.getEntityLiving();
+        }
+    }
+
+    private void resetDamage() {
+        this.damageAmount = 0;
+        this.damageSource = null;
+        this.entity = null;
+        this.hasDied = false;
+    }
+
+    @Override
+    public void writeObservationsToJSON(JsonObject json, MissionInit missionInit) {
+        EntityPlayerSP player = Minecraft.getMinecraft().player;
+        JsonObject damage_json = new JsonObject();
+        json.addProperty("is_dead", player.isDead);
+        json.addProperty("living_death_event_fired", this.hasDied);
+
+        // Note to user that there could be inaccuracies here
+        if (this.damageSourceReplaced) {
+            damage_json.addProperty("stale", true);
+            this.damageSourceReplaced = false;
```

Comment:
suggestion: Probably makes more sense to put this reinitialization assignment inside `resetDamage` along with the other reinitializations

Reply:
ahh great point